<!-- Launch Script for multi-agent ACOMMS experiments
    John P Morrison
    MIT/WHOI Joint Program
    JUL 2025

    FUNCTIONAL - USED IN JUN2025 FIELD TESTS
    - A very simple single-agent node, pings landmarks alternately with no TDMA
    - Collects range and odometry data, but doesn't try to do inter-agent comms or estimation
-->

<launch>
    <!-- Args that we would change at agent-to-agent level-->
    <arg name="agent_name" default="actor_0" doc="format: actor_0 or beacon_0"/>
    <arg name="source_addr" default="0" doc="int"/>
    <arg name="num_agents" default="1" />

    <!-- Set the ROSCONSOLE_FORMAT -->
    <env name="ROSCONSOLE_FORMAT" value="[${severity}] [${time}] [${node}]: ${message}"/>

    <!-- Start the agent node -->
    <group ns="$(arg agent_name)">

        <!-- Shared Parameters -->
        <param name="agent_name" value="$(arg agent_name)" />
        <param name="modem_address" value="$(arg source_addr)" />
        <param name="num_agents" value="$(arg num_agents)" />
        <param name="num_landmarks" value="2" />
        <rosparam param="landmarks">
            L0: [-74.5193539608157, -38.9298973079931, -1.83]
            L1: [66.5150726324041, 25.969767675496275, -1.83]
        </rosparam>

        <!-- Modem group -->
        <group ns="modem">
            <node name="acomms_driver" pkg="ros_acomms" type="acomms_driver_node.py" respawn="true" respawn_delay="10" output="log">
                <param name="modem_connection_type" value="serial" />
                <param name="modem_serial_port" value="/dev/ttyUSB0" type="str" />
                <param name="modem_baud_rate" value="19200" />
                <param name="set_modem_time" value="true" />
                <param name="modem_config/SRC" value="$(arg source_addr)" />
                <rosparam param="modem_config">
                    ALL: 0
                    BND: 0
                    MOD: 1
                    CST: 1
                    FC0: 25000
                    BW0: 5000
                    XST: 1
                    TAT: 50
                    pwramp.txlevel: 0
                    psk.packet.mod_hdr_version: 1
                </rosparam>
            </node>
        </group>

        <!-- Launch communication and graph managers-->
        <node name="comms_lbl" pkg="spurdog_acomms" type="comms_lbl.py" output="screen">
            <param name="sound_speed" value="1501" />
            <param name="sigma_range" value="1" />
            <param name="ping_timeout" value="4" />
        </node>
        <node name="odometry" pkg="spurdog_acomms" type="odometry" output="screen">
            <!-- <param name="gyro_noise_sigma" value="0.001"/>
            <param name="gyro_bias_rw_sigma" value="0.01"/> -->
            <param name="gyro_bias_x" value="0.0"/>
            <param name="gyro_bias_y" value="0.0"/>
            <param name="gyro_bias_z" value="0.0"/>
            <param name="velocity_source" value="dvl"/>
        </node>


        <!-- Launch the MOOS interface bridge -->
        <node name="udp2ros" pkg="spurdog_moos_bridge" type="udp2ros" output="log" required="true"/>

        <node name="rosbag" pkg="rosbag" type="record"
            args='-a --exclude="/$(arg agent_name)/modem/carev|/$(arg agent_name)/modem/sst|/$(arg agent_name)/modem/tta" -o /ros/logs/$(arg agent_name).bag'
            output="screen" />
    </group>
</launch>